Bring back session upload event

quinqu opened the pull request:
## Purpose 

Restores emitting `session.upload` event for S3 and disk based events. 

Fixes #5088

## Testing 
Local Log: 
`{"ei":2147483647,"event":"session.upload","code":"T2005I","time":"2021-02-03T17:13:45.46Z","cluster_name":"quin-ThinkPad-P53-auth","sid":"660228a7-ae1c-4c2d-89a9-abbe8bc15f0b","url":"file://localcluster/auth/data/log/records/multi/660228a7-ae1c-4c2d-89a9-abbe8bc15f0b"}`

External Log: 
`{"ei":2147483647,"event":"session.upload","code":"T2005I","time":"2021-02-03T17:13:00.766Z","cluster_name":"cluster","sid":"603223d8-1461-4005-8475-8ff63093a42d","url":"s3://janequin.gravitational.io/603223d8-1461-4005-8475-8ff63093a42d"}`

And verified the url at which the session upload is at exists

quinqu committed:
emit session upload event

quinqu committed:
make only protostreamer implement getuploadmetadata

quinqu committed:
go formatting

a-palchikov commented on api/types/events/events.proto:
```diff
@@ -1232,3 +1233,19 @@ message StreamStatus {
     google.protobuf.Timestamp LastUploadTime = 3
         [ (gogoproto.stdtime) = true, (gogoproto.nullable) = false ];
 }
+// SessionUpload is a session upload
```
nit: let's insert an empty line to keep this consistent with rest of the file

a-palchikov commented on lib/auth/clt.go:
```diff
@@ -2151,6 +2151,11 @@ func (c *Client) CreateAuditStream(ctx context.Context, sid session.ID) (events.
 }
 
+// GetUploadMetadata gets session upload metadata
+func (c *Client) GetUploadMetadata(sid session.ID) *events.UploadMetadata {
```
What do you need with this? I don't think it's necessary - it's an internal API

a-palchikov commented on lib/auth/apiserver.go:
```diff
@@ -55,6 +55,8 @@ type APIConfig struct {
 	// KeepAliveCount specifies amount of missed keep alives
 	// to wait for until declaring connection as broken
 	KeepAliveCount int
+	// metadataGetter gets upload metadata
```
```suggestion
	// MetadataGetter retrieves additional metadata about session uploads.
	// Will be nil if audit logging is not enabled.
```

a-palchikov commented on lib/events/api.go:
```diff
@@ -489,6 +489,22 @@ type MultipartUploader interface {
 	// earlier uploads returned first
 	ListUploads(ctx context.Context) ([]StreamUpload, error)
+	// GetUploadMetadata gets the upload metadata
+	GetUploadMetadata(sessionID session.ID) *UploadMetadata
```
This does not have to return a pointer.

a-palchikov commented on lib/events/types.go:
```diff
@@ -60,6 +60,7 @@ type (
 	SessionStart                    = events.SessionStart
 	SessionJoin                     = events.SessionJoin
 	SessionPrint                    = events.SessionPrint
+	SessionUpload                   = events.SessionUpload
```
This is not necessary - since this is new code, you can use the `api/types/events` package directly.

a-palchikov commented on lib/events/types.go:
```diff
@@ -106,6 +107,7 @@ type (
 	OneOf_UserDelete                = events.OneOf_UserDelete                //nolint
 	OneOf_UserPasswordChange        = events.OneOf_UserPasswordChange        //nolint
 	OneOf_SessionStart              = events.OneOf_SessionStart              //nolint
+	OneOf_SessionUpload             = events.OneOf_SessionUpload             //nolint
```
ditto here

quinqu committed:
remove pointer from GetUploadMetadata in MultipartUploader interface + PR feedback

pass uploadmetadata by value

update test

use DiscardAuditLog in tests

update e and use uploadHandler as UploadMetadataGetter

a-palchikov commented on lib/auth/grpcserver.go:
```diff
@@ -169,6 +171,31 @@ func (g *GRPCServer) CreateAuditStream(stream proto.AuthService_CreateAuditStrea
+			if err != nil {
+				return trail.ToGRPC(err)
+			}
+			h, err := auth.GetClusterName()
```
why h? `clusterName`?

a-palchikov commented on lib/events/api.go:
```diff
@@ -489,6 +489,22 @@ type MultipartUploader interface {
+
+// UploadMetadataGetter gets the metadata for session upload
+type UploadMetadataGetter interface {
+	GetUploadMetadata(sid session.ID) *UploadMetadata
```
nit: this can also return by value so it matches `MultipartUploader.GetUploadMetadata` - you can look at it as if `MultipartUploader` also implemented the same interface.

a-palchikov commented on lib/events/stream.go:
```diff
@@ -161,6 +161,13 @@ func (s *ProtoStreamer) ResumeAuditStream(ctx context.Context, sid session.ID, u
+// GetUploadMetadata gets the session upload metadata
+// Implements events.UploadMetadataGetter
+func (s *ProtoStreamer) GetUploadMetadata(sid session.ID) *UploadMetadata {
+	uploadMetadata := s.cfg.Uploader.GetUploadMetadata(sid)
```
You can avoid the local variable if you changed to return by value (like MultipartUploader).

a-palchikov commented on lib/events/filesessions/fileasync_chaos_test.go:
```diff
@@ -123,6 +123,7 @@ func TestChaosUpload(t *testing.T) {
 		ScanPeriod: scanPeriod,
 		Streamer:   faultyStreamer,
 		Clock:      clock,
+		AuditLog:   &events.AuditLog{},
```
If it's not meant to be functional - you can use `event.NewDiscardAuditLog` instead.

a-palchikov commented on lib/events/stream.go:
```diff
@@ -161,6 +161,12 @@ func (s *ProtoStreamer) ResumeAuditStream(ctx context.Context, sid session.ID, u
 	})
 }
 
+// GetUploadMetadata gets the session upload metadata
```
I realized that we don't need the streamer to implement `UploadMetadataGetter` - the `UploadHandler` alone is enough.

a-palchikov commented on lib/service/service.go:
```diff
@@ -1018,6 +1018,7 @@ func (process *TeleportProcess) initAuthService() error {
 
 	var emitter events.Emitter
 	var streamer events.Streamer
+	var mdGetter events.UploadMetadataGetter
```
So, this is not necessary (see below)

a-palchikov commented on lib/service/service.go:
```diff
@@ -1044,20 +1045,21 @@ func (process *TeleportProcess) initAuthService() error {
 		if err != nil {
 			return trace.Wrap(err)
 		}
+		streamer = protoStreamer
```
Neither is this.

a-palchikov commented on lib/service/service.go:
```diff
@@ -1182,6 +1185,7 @@ func (process *TeleportProcess) initAuthService() error {
 		Authorizer:     authorizer,
 		AuditLog:       process.auditLog,
 		Emitter:        checkingEmitter,
+		MetadataGetter: mdGetter,
```
Let's just use `uploadHandler` here.

a-palchikov commented:
Also, I did not know `audit_emit_failed_events` was already registered [here](https://github.com/gravitational/teleport/blob/quin/sessionUploadRebase/lib/events/auditlog.go#L104) - we should drop the change in e/ since it cause the panic due to [duplicate collector](https://drone.gravitational.io/gravitational/teleport/5919/1/6)

russjones commented:
@fspmarshall Do you mind reviewing?

fspmarshall commented on lib/events/filesessions/fileasync.go:
```diff
@@ -514,7 +521,7 @@ func (u *Uploader) upload(up *upload) error {
 		}
 	}
 
-	if err := stream.Complete(u.ctx); err != nil {
+	if err = stream.Complete(u.ctx); err != nil {
```
Nit: this can be kept as `if err := ...`.  Variable bindings within `if` blocks are allowed to shadow variable bindings in the outer scope.  This is almost always preferable since it means that the code won't need to be changed in the future if the variables in the outer scope are changed down the road.

fspmarshall commented on lib/events/recorder.go:
```diff
@@ -142,6 +142,10 @@ func (r *ForwardRecorder) GetAuditLog() IAuditLog {
 	return r.AuditLog
 }
 
+func (r *Forwarder) EmitAuditEvent(ctx context.Context, event AuditEvent) error {
+	return nil
+}
+
```
Add comments describing why this method is non-functional.  Methods which expect never to be called should return `NotImplemented` errors.

fspmarshall commented on lib/events/auditlog.go:
```diff
@@ -35,7 +35,6 @@ import (
 	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/session"
 	"github.com/gravitational/teleport/lib/utils"
-
 	"github.com/gravitational/trace"
```
Nit: we generally leave a blank line between the `gravitational/teleport` imports and imports from other repos to help visually identify external dependencies at a glance.

fspmarshall commented on lib/events/auditwriter.go:
```diff
@@ -457,7 +457,8 @@ func (a *AuditWriter) closeStream(stream Stream) {
 	ctx, cancel := context.WithTimeout(a.cfg.Context, defaults.NetworkBackoffDuration)
 	defer cancel()
-	if err := stream.Complete(ctx); err != nil {
+	err := stream.Complete(ctx)
+	if err != nil {
```
Nit: Prefer using the `if err := ...; err != nil { ... }` pattern where possible.  Keeps code more concise and is a common enough pattern to be immediately visually recognizable.

fspmarshall reviewed (commented):
Just to confirm my understanding, the reason why `SessionUpload` is emitted both in the complete step of the `GRPCServer.CreateAuditStream` method and in the `UploadCompleter.CheckUploads` method is because they two things are mutually exclusive, right?  Either the remote entity completes the upload itself, or the auth server completes the upload automatically after some timeout?

quinqu commented:
Correct @fspmarshall

quinqu committed:
feedback changes

lint

quinqu committed:
lint

quinqu committed:
remove file

russjones reviewed (approved):
Bot.

a-palchikov commented on q.sh:
```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
```
What are these for?

quinqu replied:
These actually weren't supposed to be merged, this was for my use.
